packages/ckb-sdk-core/src/signWitnesses.ts

```diff
@@ -46,30 +53,59 @@ const signWitnesses: SignWitnesses = (key: SignatureProvider | Map<LockHash, Sig
   if (!witnesses.length) throw new Error('Witnesses is empty')
 
   if (isMap(key)) {
+    if (!inputCells.length) {
+      throw new Error(`Cell shouldn't be empty when key is Map`)
+    }
     const rawWitnesses = witnesses
     const restWitnesses = witnesses.slice(inputCells.length)
     const groupedScripts = groupScripts(inputCells)
-
-    groupedScripts.forEach((indices, lockhash) => {
+    const lockhashes = [...groupedScripts.keys()]
+    for (let i = 0; i < lockhashes.length; i++) {
+      const lockhash = lockhashes[i];
       const sk = key.get(lockhash)
       if (!sk) {
         if (!skipMissingKeys) {
           throw new Error(`The signature provider to sign lockhash ${lockhash} is not found`)
         } else {
-          return
+          continue
         }
       }
 
+      const indices = groupedScripts.get(lockhash)!
       const ws = [...indices.map(idx => witnesses[idx]), ...restWitnesses]
-
-      const witnessIncludeSignature = signWitnessGroup(sk, transactionHash, ws)[0]
-      rawWitnesses[indices[0]] = witnessIncludeSignature
-    })
+      let signStatus = SignStatus.Signed
+      if (typeof sk === 'object') {
+        const witnessIncludeSignature = signWitnessGroup(sk.sk, transactionHash, ws, sk.config)[0]
+        // is multisig sign
+        const firstWitness = rawWitnesses[indices[0]]
+        if (typeof firstWitness !== 'object') {
+          throw new Error('The first witness in the group should be type of WitnessArgs')
+        }
+        let lockAfterSign = (witnessIncludeSignature as CKBComponents.WitnessArgs).lock
+        if (firstWitness.lock) {
+          lockAfterSign = firstWitness.lock + lockAfterSign?.slice(2)
+        } else {
+          lockAfterSign = serizeMultisigConfig(sk.config) + lockAfterSign?.slice(2)
+        }
+        const firstWitSigned = { ...firstWitness, lock: lockAfterSign }
+        rawWitnesses[indices[0]] = firstWitSigned
+        signStatus = getMultisigStatus(sk.config, [...sk.signatures, sk.blake160])
+      } else {
+        const witnessIncludeSignature = signWitnessGroup(sk, transactionHash, ws)[0]
+        rawWitnesses[indices[0]] = witnessIncludeSignature
+      }
+      if (signStatus === SignStatus.Signed) {
+        indices.forEach(idx => {
+          const wit = rawWitnesses[idx]
+          rawWitnesses[idx] = typeof wit === 'string' ? wit : serializeWitnessArgs(wit)
+        })
+      }
+    }
     return rawWitnesses
   }
 
   const signedWitnesses = signWitnessGroup(key, transactionHash, witnesses)
-  return signedWitnesses
+  return signedWitnesses.map(wit => typeof wit === 'string' ? wit : serializeWitnessArgs(wit))
```

Comment:
Serializing all witnesses makes intact witnesses unavailable in the next round of signing.

For example, witnesses `[A, B, C]` was signed by private keys `[pa, pc]` in the first round while `skipMissingKeys = true`, partially signed witnesses `[signedA, unsignedB, signedC]` can still be signed by `[pb]`. But if `[unsignedB]` is serialized in the first round, then `[signedA, unsignedB, signedC]` cannot be signed anymore.

Reply:
Do you mean it will sign multi times when it's not multisig?

Reply:
Yes, a transaction could have multiple input groups(witness groups) and can be signed one by one, even they are all secp256k1/blake160 lock script, that's unusual but allowed.

There're cases allowing users to sign the same transaction respectively, e.g. https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki

Reply:
If input cells have multiple input groups, the `key` in parameters should be `Map`, it will not execute this branch.
If it executes this branch, input cells should only include one group.

Reply:
What about a transaction including witnesses `[A, B]` and signed by private keys `pa`, `pb` respectively.

Reply:
I think one group should be only signed once. During one execution, only the same group of witnesses will be serialized.
If first signed A, witness B will not be serialized.

Reply:
If a transaction includes witnesses [A, B], it may be called like this.
1. it should be called with `key` that is a `Map`
2. called with different witnesses two times when `key` it not a `Map`

Because `signWitnessGroup` use `witnessGroup[0]` as the first witness.

Reply:
> Do you mean it will be called more times when key is not Map? Then I think I should not serialize all witnesses.

Yes, it is possible to sign a transaction with private key `pa` and private key `pb` separately, and only one key is used each time.

The most classical case is open transaction. For example, user A wants to swap 100 ckb and 100 wckb with user B, he signed a transaction that send 100 ckb to user B with a requirement of receiving 100wckb from user B, then send the partially signed transaction to user B and user B should sign it to make the whole transaction valid.

In this case, witnesses of user B's input group should be intact.